### PR TITLE
Always return a cycle in `digraph_find_cycle` if no node is specified and a cycle exists

### LINKED ISSUE #1171: rx.digraph_find_cycle() does not find the cycle in a graph
### Information

- **rustworkx version**: 0.14.2
- **Python version**: 3.10.14
- **Rust version**: rustc 1.66.0-nightly (8b705839c 2022-09-26)
- **Operating system**: Linux

### What is the current behavior?
`rx.is_directed_acyclic_graph(graph)` returns False but `rx.digraph_find_cycle(graph)` fails to find the cycle in the graph.

### What is the expected behavior?
When `rx.is_directed_acyclic_graph(graph)` returns False, `rx.digraph_find_cycle(graph)` should find a cycle in the graph.

### Steps to reproduce the problem
Running the following script several times. It quite often reproduces the bug. The graph is always correctly identified as not a DAG, but quite often `rx.digraph_find_cycle(graph)` does not find a cycle.
```
import rustworkx as rx

edges = [
    (6, 7),
    (8, 9),
    (15, 16),
    (0, 6),
    (2, 8),
    (9, 10),
    (7, 10),
    (10, 11),
    (11, 12),
    (10, 12),
    (12, 13),
    (12, 14),
    (4, 15),
    (16, 17),
    (14, 3),
    (13, 17),
    (17, 18),
    (18, 19),
    (17, 19),
    (19, 20),
    (19, 5),
    (20, 1),
    (27, 28),
    (28, 29),
    (34, 35),
    (21, 27),
    (23, 29),
    (29, 30),
    (30, 31),
    (29, 31),
    (31, 32),
    (31, 33),
    (25, 34),
    (32, 24),
    (35, 36),
    (33, 36),
    (36, 37),
    (37, 38),
    (36, 38),
    (38, 39),
    (38, 26),
    (39, 22),
    (42, 45),
    (44, 45),
    (40, 44),
    (45, 46),
    (46, 47),
    (45, 47),
    (47, 48),
    (47, 49),
    (48, 43),
    (49, 41),
    (56, 57),
    (57, 58),
    (61, 63),
    (50, 56),
    (52, 57),
    (58, 59),
    (57, 59),
    (59, 60),
    (54, 62),
    (59, 61),
    (60, 53),
    (62, 63),
    (63, 64),
    (64, 65),
    (63, 65),
    (65, 66),
    (65, 67),
    (66, 51),
    (67, 55),
    (74, 75),
    (75, 76),
    (80, 81),
    (68, 74),
    (70, 76),
    (76, 77),
    (77, 78),
    (76, 78),
    (78, 79),
    (78, 80),
    (72, 81),
    (79, 71),
    (81, 82),
    (82, 83),
    (81, 83),
    (83, 84),
    (83, 69),
    (84, 73),
    (1, 42),
    (3, 23),
    (24, 72),
    (22, 50),
    (5, 54),
    (41, 40),
    (26, 52),
    (69, 70),
]
edges = [x + (None,) for x in edges]
nodes = set()
for edge in edges:
    nodes.add(edge[0])
    nodes.add(edge[1])
nodes = list(nodes)

# Create a directed graph (PyDiGraph)
graph = rx.PyDiGraph()

# Add nodes and edges
graph.add_nodes_from(nodes)
graph.add_edges_from(edges)

# Check if the graph is a DAG
is_dag = rx.is_directed_acyclic_graph(graph)
print("Is the graph a DAG? {}".format(is_dag))

# Attempt to find a cycle
cycle = rx.digraph_find_cycle(graph)
print("Cycle found : {}".format(cycle))
```
Quite often the script produces:
```
Is the graph a DAG? False
Cycle found : EdgeList[]
```

### PR DESCRIPTION
**Pull request by IvanIsCoding**
Closes #1171 

Instead of picking an arbitrary node if it is not provided, we find a smarter pick that is in a strongly-connected component with more than one node. It also handles the case of self-loops in case there is a SCC with a single node.

### CODE REVIEW

File: rustworkx-core/src/connectivity/find_cycle.rs [hunk 1]
```diff
@@ -54,25 +56,30 @@ use std::hash::Hash;
 /// ```
 pub fn find_cycle<G>(graph: G, source: Option<G::NodeId>) -> Vec<(G::NodeId, G::NodeId)>
 where
+    G: Copy,
```
Comment (mtreinish) [inline rustworkx-core/src/connectivity/find_cycle.rs:59]: This seems like a pretty big API compatibility issue. I wouldn't expect most graph types to by `Copy`. `Clone` sure, but `Copy` seems to go against the spirit of the trait to use if for graph objects.

File: releasenotes/notes/fix-digraph-find-cycle-141e302ff4a8fcd4.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+---
+fixes:
```
Comment (mtreinish) [inline releasenotes/notes/fix-digraph-find-cycle-141e302ff4a8fcd4.yaml:2]: I think we need an upgrade note about the trait bound changes to the public rustworkx-core function.

File: rustworkx-core/src/connectivity/find_cycle.rs [hunk 3]
```diff
@@ -119,11 +126,49 @@ where
...
+    G: EdgeCount,
+    for<'b> &'b G: GraphBase<NodeId = G::NodeId>
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + Visitable
+        + IntoEdgeReferences
+        + NodeIndexable,
+    G::NodeId: Eq + Hash,
+{
+    for scc in algo::tarjan_scc(&graph) {
```
Comment (mtreinish) [inline rustworkx-core/src/connectivity/find_cycle.rs:142]: I remember we had panics or some issue with one of petgraph's strongly connected component functions along time ago. Was it kosaraju or tarjan?

File: rustworkx-core/src/connectivity/find_cycle.rs [hunk 4]
```diff
@@ -54,25 +56,30 @@ use std::hash::Hash;
...
     G: NodeCount,
     G: EdgeCount,
-    for<'b> &'b G: GraphBase<NodeId = G::NodeId> + IntoNodeIdentifiers + IntoNeighborsDirected,
+    for<'b> &'b G: GraphBase<NodeId = G::NodeId>
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + Visitable
+        + NodeIndexable
+        + IntoEdgeReferences,
     G::NodeId: Eq + Hash,
```
Comment (mtreinish) [inline rustworkx-core/src/connectivity/find_cycle.rs:69]: This is a potential API breakage because we're requiring additional traits on the `graph` input.

**Review by mtreinish — COMMENTED** [review]
Overall I like the idea here it makes sense to me to use a connected component function to get a node in a cycle and go from there. I just have some concerns about the backwards compatibility implications. They're likely unavoidable to use the scc functions, but we should document it as an upgrade change then.

File: rustworkx-core/src/connectivity/find_cycle.rs (hunk 1 above)
Comment (IvanIsCoding) [reply rustworkx-core/src/connectivity/find_cycle.rs]: I fixed it, with `kosaraju_scc` and references I don't need copy anymore

File: rustworkx-core/src/connectivity/find_cycle.rs (hunk 3 above)
Comment (IvanIsCoding) [reply rustworkx-core/src/connectivity/find_cycle.rs]: Switched to `kosaraju-scc` which is the one we use for the Python SCC function

File: rustworkx-core/src/connectivity/find_cycle.rs (hunk 4 above)
Comment (IvanIsCoding) [reply rustworkx-core/src/connectivity/find_cycle.rs]: I got rid of all traits except `Visitable`

File: releasenotes/notes/fix-digraph-find-cycle-141e302ff4a8fcd4.yaml (hunk 2 above)
Comment (IvanIsCoding) [reply releasenotes/notes/fix-digraph-find-cycle-141e302ff4a8fcd4.yaml:2]: Done

**Comment by IvanIsCoding** [comment]
> Overall I like the idea here it makes sense to me to use a connected component function to get a node in a cycle and go from there. I just have some concerns about the backwards compatibility implications. They're likely unavoidable to use the scc functions, but we should document it as an upgrade change then.

I removed most of the traits but `petgraph::visit::Visitable` is a requirement for `kosaraju_scc` and by https://docs.rs/petgraph/latest/petgraph/visit/trait.Visitable.html most graphs implement that

**Review by mtreinish — APPROVED** [review]
LGTM now, thanks for the quick update.

**Pull request merged**